pyro/ops/tensor_utils.py

```diff
@@ -13,3 +13,30 @@ def block_diag(m):
     B, M, N = m.shape
     eye = torch.eye(B, dtype=m.dtype, device=m.device).reshape(B, 1, B, 1)
     return (m.unsqueeze(-2) * eye).reshape(B * M, B * N)
+
+
+def _complex_mul(a, b):
+    ar, ai = a.unbind(-1)
+    br, bi = b.unbind(-1)
+    # FIXME should b be conjugated?
+    # bi = -bi  # Conjugates b.
+    return torch.stack([ar * br - ai * bi, ar * bi + ai * br], dim=-1)
+
+
+def conv1d_fft(signal, kernel):
+    """
+    Computes the 1-d convolution of signal by kernel using FFTs.
+    The two arguments should have the same rightmost dim, but may otherwise be
+    arbitrarily broadcastable.
+
+    :param torch.Tensor signal: A signal to convolve.
+    :param torch.Tensor kernel: A convolution kernel.
+    :return: A tensor of shape
+        ``broadcast_shape(convolve, kernel.shape)``
+    :rtype torch.Tensor:
+    """
+    assert signal.size(-1) == kernel.size(-1)
+    f_signal = torch.rfft(torch.cat([signal, torch.zeros_like(signal)], dim=-1), 1)
```

Comment:
This is relevant to [autocorrelation](https://github.com/pyro-ppl/pyro/blob/dev/pyro/ops/stats.py#L95). I think [next_fast_len](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fftpack.next_fast_len.html) might be helpful for performance. Let me verify tomorrow if we need `onesided` arg  in `rfft` and `irfft`.

Reply:
done, I've updated padding logic.